tests/Unit/Evolution/Systems/ScalarWave/Test_Characteristics.cpp

```diff
@@ -0,0 +1,388 @@
+#include "PointwiseFunctions/MathFunctions/MathFunction.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "tests/Unit/Pypp/CheckWithRandomValues.hpp"
+#include "tests/Unit/Pypp/SetupLocalPythonEnvironment.hpp"
+
+// IWYU pragma: no_forward_declare Tags::dt
+// IWYU pragma: no_forward_declare Tensor
+// IWYU pragma: no_forward_declare Variables
+
+namespace {
+template <size_t Index, size_t Dim>
+Scalar<DataVector> speed_with_index(
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal) {
+  return Scalar<DataVector>{
+      ScalarWave::CharacteristicSpeedsCompute<Dim>::function(normal)[Index]};
+}
+
+template <size_t Dim>
+void test_characteristic_speeds() noexcept {
+  const DataVector used_for_size(5);
+  pypp::check_with_random_values<1>(speed_with_index<0, Dim>, "Characteristics",
+                                    "char_speed_upsi", {{{-10.0, 10.0}}},
+                                    used_for_size);
+  pypp::check_with_random_values<1>(speed_with_index<1, Dim>, "Characteristics",
+                                    "char_speed_uzero", {{{-10.0, 10.0}}},
+                                    used_for_size);
+  pypp::check_with_random_values<1>(speed_with_index<3, Dim>, "Characteristics",
+                                    "char_speed_uminus", {{{-10.0, 10.0}}},
+                                    used_for_size);
+  pypp::check_with_random_values<1>(speed_with_index<2, Dim>, "Characteristics",
+                                    "char_speed_uplus", {{{-10.0, 10.0}}},
+                                    used_for_size);
+}
+
+// Test return-by-reference char speeds by comparing to analytic solution
+template <size_t Dim>
+void test_characteristic_speeds_analytic(
+    const size_t grid_size_each_dimension) noexcept {
+  // Setup mesh
+  Mesh<Dim> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+                 Spectral::Quadrature::GaussLobatto};
+  const auto x_logical = logical_coordinates(mesh);
+  // Get ingredients
+  const size_t n_pts = x_logical.begin()->size();
+  // Outward 3-normal to the surface on which characteristic fields are needed
+  tnsr::i<DataVector, Dim, Frame::Inertial> unit_normal_one_form{
+      DataVector(n_pts, 1.)};
+  const auto norm_of_one_form = magnitude(unit_normal_one_form);
+  for (size_t i = 0; i < Dim; ++i) {
+    unit_normal_one_form.get(i) /= get(norm_of_one_form);
+  }
+
+  // Get characteristic speeds locally
+  const auto upsi_speed =
+      make_with_value<Scalar<DataVector>>(unit_normal_one_form, 0.);
+  const auto uzero_speed =
+      make_with_value<Scalar<DataVector>>(unit_normal_one_form, 0.);
+  const auto uplus_speed =
+      make_with_value<Scalar<DataVector>>(unit_normal_one_form, 1.);
+  const auto uminus_speed =
+      make_with_value<Scalar<DataVector>>(unit_normal_one_form, -1.);
+
+  // Check that locally computed fields match returned ones
+  const auto char_speeds_from_func =
```

Comment:
Here (and below) you could drop the `_from_func` suffix, after renaming the expected results above

Reply:
Made this change here and everywhere else